### ISSUE TITLE
rerun autoreconf in 1.4.1 release tarball

### ISSUE THREAD

**Issue by dpryan79** [samtools/samtools#679]
@jenniferliddle can you rerun `autoreconf` in the htslib subdirectory of the 1.4.1 release tar ball? The current `htslib/configure` is broken in it.

**Comment by dpryan79** [comment]
As an example, here are the first 5 lines of what I downloaded:

    #! /bin/sh
    # Guess values for system-dependent variables and create Makefiles.
    # Generated by GNU Autoconf 2.68 for HTSlib make1: Entering directory `/nfs/users/nfs_j/js10/tmp/dist/c-maint/htslib-1.4.1'
    1.4.1
    make1: Leaving directory `/nfs/users/nfs_j/js10/tmp/dist/c-maint/htslib-1.4.1'.

So stderr or stdout got mixed in.

**Comment by jenniferliddle** [comment]
Thanks for the report, I'll take a look...

**Comment by dpryan79** [comment]
@jenniferliddle Great, thanks!

**Comment by dpryan79** [comment]
The same issue exists in the htslib-1.4.1 release tar ball, which I guess shouldn't be surprising.

**Comment by dpryan79** [comment]
Cf. this [Travis-CI](https://travis-ci.org/bioconda/bioconda-recipes/jobs/229957108#L754-L756) error.

**Comment by jenniferliddle** [comment]
I have updated the tarballs for htslib, samtools, and bcftools on both github and sourceforge. Hopefully this will solve the problems.
Thanks for reporting it so quickly.

**Comment by jkbonfield** [comment]
I'm trying to work out what caused this, but my suspicions are they come from this rather bizarre line in the configure.ac script:

```
dnl Process this file with autoconf to produce a configure script
AC_INIT([Samtools], m4_esyscmd_s([make print-version]),
[samtools-help@lists.sourceforge.net], [], [http://www.htslib.org/])
```
https://github.com/samtools/samtools/blob/develop/configure.ac#L26

(Autoconf depending on execution of make to generate the configure script is rather *interesting*!  I fixed this for samtools at least in my windows port, but clearly need to do on other repositories too.)

Long term I'd like to throw away the c-maint repository and replace it with good old "make dist" and "make distcheck" targets so we know things actually work without requiring manual intervention to test the release process.  I think it's quite straight forward to do so I'll take a look at it.

**Comment by dpryan79** [comment]
[Still broken](9ff870a8d7554875246c47c6192e568a), also on the command line directly.

**Comment by dpryan79** [comment]
BTW, I have samtools, htslib and bcftools in the PR to bioconda, so they're all getting tested at once.

**Comment by jmarshall** [comment]
> Autoconf depending on execution of make to generate the configure script is rather interesting! 

Yes, it's a lovely hack isn't it :smile:.  It's non-trivial to get the right version number into configure otherwise.

Unfortunately it's not without problems; cf [Debian's patch](https://anonscm.debian.org/cgit/debian-med/samtools.git/tree/debian/patches/literal_version.patch?h=debian/1.3.1-3), though that solution only works for releases, not arbitrary possibly-dirty Git working directories.  Whether having the correct version number burnt into _configure_ is a priority is another question, though it does appear in `--help` output and _config.status_.

**Comment by jkbonfield** [comment]
I explored just adding ```git describe``` in there, but that only works on git clones so it's not a good general purpose solution.  It probably needs an m4 macro with the full script in there to check for .git first.

Looking elsewhere I see the usual route is to add a version.sh or similarly named shell script, which also works OK but maybe overkill unless our script gets complex.

**Comment by jenniferliddle** [comment]
I have updated the tarballs for htslib, samtools, and bcftools on both github and sourceforge. 
Again.
Hopefully *this* will solve the problems.

**Comment by dpryan79** [comment]
Works now, thanks!

**Comment by jmarshall** [comment]
> Long term I'd like to throw away the c-maint repository and replace it with good old "make dist"

Even invoked as `make dist` it's still just a few shell commands.  Modulo some differently-rough edges, I would think the scripts in c-maint are largely what you would write in `make dist` as they are what you get when you extrapolate (1) not duplicating the release script over the three projects; and more importantly (2) injecting the htslib source tree into the other two release tarballs.

**Comment by jkbonfield** [comment]
Perhaps but

(1) It's not really much work to do and duplication isn't necessarily a bad thing if it's what people expect.  The same argument could be applied to all the other duplicated parts of our build setup too.  It just feels having yet another repository is a lot of extra fluff with VERY little gain (and infact a lot of pain  it turns out due to bugs in it). 

Let's look at those scripts.  There is addhtslib which appears to be a rewrite of "autoreconf" run in the subdirectory.  A trivial one liner for a Makefile.  Then there is mktarball which also isn't that long to duplicate plus has issues if it is run under anything bar MacOS X. 

(2) Personally I dislike the whole joint tarball idea as it can lead to confusion.  Eg in the past you released 1.3.2 as an update to htslib only without updating samtools.  What would we call the joint tarball?  Is it samtools-1.3.2.tar.gz (wrong as it's not)?  Something ghastly like samtools-1.3.1-1.3.2.tar.gz? 

You avoided the whole issue by simply not updating the tarball because nothing else made much sense.  However that is the worst scenario.  We have an important update to make yet leave a trap for people to walk into by having a single "all you need is here" tarball with the known buggy code still in it.

All we're really doing by having a single tarball is saving people one extra download and one extra untar.  It's really not a lot of extra effort and if you can't manage that then frankly I don't want to be the one managing the inevitable flurry of how-do-I-make-a-compiler-work queries.  By keeping it though we're locking ourselves into having to always releasing htslib and samtools together and to having another strange corner case in an already overly complex build system that treats htslib in a subdirectory as special and different to htslib in a sibling directory.

A work in progress, but I was envisioning something like this work-in-progress in the Makefile (and to throw away the huge pitfall of "distsrc", which did a very good job of trashing my home directory yesterday!):

```
DDIR=htslib-$(PACKAGE_VERSION)
DTAR=$(DDIR).tar.gz

# NB: make sure you run this on a committed branch as uncommitted changes                                       
# will not be copied.                                                                                           
dist:
        -rm -rf _dist
        -mkdir -p _dist/$(DDIR)
        if [ -e .git ]; then git archive HEAD . | (cd _dist/$(DDIR) && tar xf - && autoreconf); else cp -r *.[c\
h15] [ILMNRcht]*[ELSbcekmnt] _dist/$(DDIR); fi
        (cd _dist; tar cfz ../$(DTAR) $(DDIR))
        rm -rf _dist                                                                                           

# Make and extract the distribution tar ball followed by configure, make                                        
# and check.  We then rebuild the tar ball again to validate it matches                                         
# the original one.  This validates our make dist can cope with debris                                          
# left behind from a build.                                                                                     
distcheck: dist
        -mkdir _dist
        cd _dist && tar xf ../$(DTAR)
        cd _dist/htslib-* && ./configure $${CONFIG_OPTS:-} && $(MAKE) && $(MAKE) check
        cd _dist/htslib-* && $(MAKE) dist
        tar tvf $(DTAR) |awk '{sub("^[^/]*/","",$$6);print $$3,$$6}' > _dist/curr.lst
        tar tvf _dist/htslib-*/htslib-*.tar.gz |awk '{sub("^[^/]*/","",$$6);print $$3,$$6}' > _dist/sub.lst
        diff _dist/*.lst
        -rm -rf _dist
```

The above still needs some fixing, specifically tidying up of autom4te.cache as this turns out to be the cause of the rogue make warning, but I think the print-version stuff should be replaced too which would solve that issue as well.

**Comment by jmarshall** [comment]
It's great that you all are revisiting this, and I don't think there's much to be gained from me discussing it other than to suggest thinking before just “throw[ing] away” what other maintainers did, as there are reasons why all those things accreted.  (Indeed if there are issues in the existing scripts, you might consider fixing them.)

The joint tarballs are of course a response to the end users, who just want to get **samtools** or **bcftools** installed and for the most part have less than zero interest in htslib-related implementation details.

> Eg in the past you released 1.3.2 as an update to htslib only without updating samtools. […] You avoided the whole issue by simply not updating the tarball because nothing else made much sense.

This is inaccurate, James.  That was a decision that the samtools maintainers came to as a group, not something that _I_ did.

**Comment by jkbonfield** [comment]
Rest assured we are thinking before just throwing away what past maintainers did, and the git logs are there to explain things (mostly).  However we wouldn't even be looking at this if the existing implementation worked well, but it doesn't - two releases and two different problems with a 25 line script implies it's due a rethink.   Extra complexity such as an entirely new repository for holding a couple dozen lines of code also seems over the top.

Given we are working on it then moving to a more standard process  ("make dist" and "make distcheck" are standard) seems to me to be the obvious move and makes it easier for new maintainers.  You're welcome to comment on the pull request when it lands, to double check that we didn't miss out some corner case.
